borg/repository.py

```diff
@@ -591,6 +602,28 @@ def get_fd(self, segment):
             self.fds[segment] = fd
             return fd
 
+    def _sync_and_close(self, fd):
+        fd.flush()
+        os.fdatasync(fd.fileno())
+        fd.close()
+
+    def close_segment(self, sync=False):
+        if len(self._write_fds) == self.MAX_PENDING_WRITE_FDS:
+            self._sync_and_close(self._write_fds.popleft())
+        if self._write_fd:
+            self.segment += 1
+            self.offset = 0
+            self._write_fds.append(self._write_fd)
+            if hasattr(os, 'posix_fadvise'):  # only on UNIX
+                # tell the OS that it does not need to cache what we just wrote,
+                # avoids spoiling the cache for the OS and other processes.
+                os.posix_fadvise(self._write_fd.fileno(), 0, 0, os.POSIX_FADV_DONTNEED)
+            self._write_fd = None
+        if sync:
+            for fd in self._write_fds:
+                self._sync_and_close(fd)
+            self._write_fds.clear()
```

Comment:
Ah yes. If the commit was written() before the fsync()s then (at least hypothetically) the OS could write the commit to disk first.